wingtips-servlet-api/src/main/java/com/nike/wingtips/servlet/RequestTracingFilter.java

```diff
@@ -1,121 +1,199 @@
-                                                                   TracingState originalRequestTracingState) {
-        // Async processing was started, so we have to complete it with a listener.
-        asyncRequest.getAsyncContext().addListener(
-            new WingtipsRequestSpanCompletionAsyncListener(originalRequestTracingState)
-        );
+    protected void doFilterInternal(HttpServletRequest request,
+                                    HttpServletResponse response,
+                                    FilterChain filterChain) throws ServletException, IOException {
+        // Surround the tracing filter logic with a try/finally that guarantees the original tracing and MDC info found
+        //      on the current thread at the beginning of this method is restored to this thread before this method
+        //      returns, even if the request ends up being an async request. Otherwise there's the possibility of
+        //      incorrect tracing information sticking around on this thread and potentially polluting other requests.
+        TracingState originalThreadInfo = TracingState.getCurrentThreadTracingState();
+        try {
+            // See if there's trace info in the incoming request's headers. If so it becomes the parent trace.
+            Tracer tracer = Tracer.getInstance();
+            final Span parentSpan = HttpSpanFactory.fromHttpServletRequest(request, getUserIdHeaderKeys());
+            Span newSpan;
+
+            if (parentSpan != null) {
+                logger.debug("Found parent Span {}", parentSpan);
+                newSpan = tracer.startRequestWithChildSpan(parentSpan, HttpSpanFactory.getSpanName(request));
+            } else {
+                newSpan = tracer.startRequestWithRootSpan(
+                    HttpSpanFactory.getSpanName(request),
+                    HttpSpanFactory.getUserIdFromHttpServletRequest(request, getUserIdHeaderKeys())
+                );
+                logger.debug("Parent span not found, starting a new span {}", newSpan);
+            }
+
+            // Put the new span's trace info into the request attributes.
+            request.setAttribute(TraceHeaders.TRACE_SAMPLED, newSpan.isSampleable());
+            request.setAttribute(TraceHeaders.TRACE_ID, newSpan.getTraceId());
+            request.setAttribute(TraceHeaders.SPAN_ID, newSpan.getSpanId());
+            request.setAttribute(TraceHeaders.PARENT_SPAN_ID, newSpan.getParentSpanId());
+            request.setAttribute(TraceHeaders.SPAN_NAME, newSpan.getSpanName());
+            request.setAttribute(Span.class.getName(), newSpan);
+
+            // Make sure we set the trace ID on the response header now before the response is committed (if we wait
+            //      until after the filter chain then the response might already be committed, silently preventing us
+            //      from setting the response header)
+            response.setHeader(TraceHeaders.TRACE_ID, newSpan.getTraceId());
```

Comment:
Ps eventhough b3 doesnt usually do this, I noticed amazon do. For example returning back the sample result (as everything is in one header)

Reply:
Yeah I still have issue #30 to change this to have the response header be something other than `X-B3-TraceId`, but having the trace ID somewhere in the response headers has been critically useful.